jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -1382,10 +1383,58 @@ public void demand()
         }
 
         @Override
-        public void prepareResponse(HttpFields.Mutable headers)
+        public void prepareResponse(HttpFields.Mutable responseHeaders)
         {
-            if (_connectionKeepAlive && _version == HttpVersion.HTTP_1_0 && !headers.contains(HttpHeader.CONNECTION, HttpHeaderValue.CLOSE.asString()))
-                headers.add(HttpFields.CONNECTION_KEEPALIVE);
+            boolean hasConnectionClose = responseHeaders.contains(HttpHeader.CONNECTION, HttpHeaderValue.CLOSE.asString());
+            boolean hasConnectionKeepAlive = responseHeaders.contains(HttpHeader.CONNECTION, HttpHeaderValue.KEEP_ALIVE.toString());
```

Comment:
again, don't lookup values if we don't need them, especially in this case as it involves a full iteration to do each lookup.

So lookup only if needed and perhaps look for close first, so it can override keepalive.

Reply:
The problem is that each of these could then be looked up multiple times instead.
So I took the case of the lookup occurring 4 times down to 2, which I felt was better.

Reply:
A single iteration through is best, I'll work on that after you have had a look at the `ResponseHeadersTest`.

In particular the new tests ...

* `testHTTP10ConnectionBehavior` for when the client is using HTTP/1.0
* `testHTTP11ConnectionBehavior` for when the client is using HTTP/1.1

Are you good with the set of expectations for those?
Each test case arguments have 3 parameters.

1. The `Connection` request header as sent from the Client (null means no Connection header sent)
2. The `Connection` response header as set by an application (Servlet in this case.  null means no Connection header is set)
3. The `Connection` response header as seen by the client. (null means no Connection header should be present received)

Once I get that set of tests sane, then I can look closer at the other tests that are now failing due to this PR.
Some of which are suspect (like a test client that uses HTTP/1.0 but also `Connection: close`, which is not typical, but we should accept it, and the new testcases I mentioned has variants of this scenario too)

Reply:
If you take just the changes to the `ResponseHeadersTest` back to `jetty-12.0.x` HEAD and run them, you'll quickly see the various cases that currently return a `Connection: ` (empty value) response header.

Reply:
I'm not saying avoid local variables with those looked up values (so  we don't need to lookup multiple times), but I am saying delay doing the lookup until it is known that it is needed.